### PR TITLE
OpenSea security provider metrics

### LINKED ISSUE #16262: Wallet Tx Security Provider: add a new user property and a ui_customizations property
## Background
[Epic on Notion](https://www.notion.so/Enable-MVP-support-for-tx-security-providers-954fef15373d4736b046566d20c93de1)
We want to update our metric scheme to capture the usage and efficacy of the new transactions security provider feature so we're able confidently iterate over and improve this feature.
As always, for anyone looking for more info on MetaMetrics, [here](https://medium.com/metamask/metametrics-a-refresher-on-metamasks-analytics-system-78d725cb9a75) is a good article.

## Tasks

1. Create a new user profile property (more info [here](https://help.mixpanel.com/hc/en-us/articles/115004501966#user-profile-properties)) called `security_provider`. This property should be a type list property (more info [here](https://help.mixpanel.com/hc/en-us/articles/115004547063-Properties-Supported-Data-Types#list)) and receive the value `opensea` if the user has enabled the transactions security provider feature under Settings > Experimental. Related ticket here https://github.com/MetaMask/metamask-extension/issues/16257.

2. If the user turns off the transactions security provider feature under Settings > Experimental, the user profile property `security_provider` should be updated to remove the `opensea` value from it.

3. Add a new property “ui_customizations” to `Transaction Approved`, `Transaction Rejected`, `Signature Approved` and `Signature Rejected` events.
This property should be a type list property (more info [here](https://help.mixpanel.com/hc/en-us/articles/115004547063-Properties-Supported-Data-Types#list)) and receive the value `flagged_as_malicious` when the transaction was flagged as malicious by the transaction security provider. Related ticket here https://github.com/MetaMask/metamask-extension/issues/16258 .

## Acceptance Criteria
1. This new properties are only collect for users that have opted in MetaMetrics.

### LINKED ISSUE #18150: Add metrics for transactions and signatures flagged as unknown for the Tx Security Provider feature
Add `ui_customizations` property with value `flagged_as_unknown_security` whenever we display the yellow warning (Security provider API response is `2`).
This is a follow up/extension of https://github.com/MetaMask/metamask-extension/pull/17688 and should follow the same spec as that one.

### PR DESCRIPTION
**Pull request by filipsekulic**
## Explanation
* Fixes #16262 
* Fixes #18150

Added metrics for the `OpenSea` security provider. 

Created a new user profile property called `security_provider`. This property is a type list and receives the value `opensea` if the user has enabled the `OpenSea` security provider feature under `Settings > Experimental`.
If the user turns off the security provider under `Settings > Experimental`, then the user profile property `security_provider` is updated and the `opensea` value removed from it.

Additionally there is a new property called `ui_customizations` added to `Transaction Approved`, `Transaction Rejected`, `Signature Approved` and `Signature Rejected` events.
This property receives one of the following values: 

1. `['flagged_as_malicious']` when the transaction was flagged as malicious
2. `['flagged_as_safety_unknown']` when the transaction was flagged as safety unknown
3. `null` in all other cases 

by the transaction security provider.

## Manual testing steps
In the `.metamaskrc` set the transaction security feature flag to 1 - `TRANSACTION_SECURITY_PROVIDER=1`.

Follow the procedure for testing the metrics event and check the response in the background console of the extension - `background.html`.

Cases to check:
- enable/disable the `OpenSea` security provider under `Settings > Experimental`
- `SEND LEGACY TRANSACTION`
- `SEND EIP 1559 TRANSACTION`
- `ETH SIGN`
- `PERSONAL SIGN`
- `SIGN TYPED DATA`
- `SIGN TYPED DATA V3`
- `SIGN TYPED DATA V4`

### CODE REVIEW

**Comment by jpuri** [comment]
Changes look good to me, it will be nice to have a review by @brad-decker

**Review by jpuri — DISMISSED** [review]

File: app/scripts/lib/message-manager.js
Comment (brad-decker) [inline app/scripts/lib/message-manager.js:254]: I believe the ui_customization should be added to the events in createRPCMethodTrackingMiddleware versus adding a new event and renaming events here and in the other message managers.

**Review by brad-decker — CHANGES_REQUESTED** [review]

**Comment by bschorchit** [comment]
A request from data team: we should use `security_providers` (plural) instead of `security_provider` (singular) for the name of the user profile property. Could you make this small change, @filipsekulic ? Thank you!

File: app/scripts/lib/message-manager.js
Comment (filipsekulic) [reply app/scripts/lib/message-manager.js:254]: Hey @brad-decker, 
Do you have a suggestion how it should look like? I mean, in the middleware you mentioned I need the information I get from the message - `msg.securityProviderResponse?.flagAsDangerous`. 
Also, I have to mention the presence of the property that shares the same name as the one mentioned above (https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/lib/createRPCMethodTrackingMiddleware.js#L173).

Comment (brad-decker) [reply app/scripts/lib/message-manager.js:254]: You would add your value to the array already present. The events in these confirmations are not the ones we rely upon. Let me dig in a bit and see how you can get access to the value you need

Comment (filipsekulic) [reply app/scripts/lib/message-manager.js:254]: @brad-decker I tried it, but without success. The reason is that the data being passed to the `securityProviderRequest` from `createRPCMethodTrackingMiddleware` (`req`) is not proper, so as the response in that case. The proper data being sent is shown in the attached image (`personal-message-manager.js`).

Data

Comment (brad-decker) [reply app/scripts/lib/message-manager.js:254]: @filipsekulic look at the detectSIWE implementation. Its done in both places as well. You will need to formulate the data object that securityProviderCheck needs -- note that in perosnal-message-manager that includes the 'id' of the transaction but the id key isn't used.

Comment (filipsekulic) [reply app/scripts/lib/message-manager.js]: @brad-decker I implemented the solution you suggested. Thanks! However, I could not fix the failing test - `createRPCMethodTrackingMiddleware.test.js`. I must be missing something. Hope you can help me. I would appreciate your help!

**Comment by seaona** [comment]
Some issues from QA:
- When I load MM for the first time, I can see a background console error related to `detectSIWE` , saying `TypeError: The first argument must be one of type string, Buffer, ArrayBuffer, Array, or Array-like Object. Received type undefined`

image

- *_Sometimes_* the event for Send/Send Legacy is not emitted. 

https://user-images.githubusercontent.com/54408225/222410134-e357bf3f-1456-4e49-a38d-d72169ef0b97.mp4


- I haven't observed any metrics event related to the provider, when interacting with Signatures (any kind of signature)

https://user-images.githubusercontent.com/54408225/222410291-d0e404ba-9e9d-4a2f-84eb-419cd709baee.mp4

File: app/scripts/lib/personal-message-manager.js
Comment (brad-decker) [inline app/scripts/lib/personal-message-manager.js:229]: This code needs to be removed, events for approval are already tracked in createRPCMethodTrackingMiddleware

File: app/scripts/lib/message-manager.js
Comment (brad-decker) [inline app/scripts/lib/message-manager.js:193]: This code needs to be removed, events for approval are already tracked in createRPCMethodTrackingMiddleware

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (brad-decker) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:151]: This code should only be fired if we are dealing with a signature request, which is what @seaona 's QA item is showing (this method is called for *every* request to the RPC handler, not just signatures, note the previous usage of detectSIWE happens ONLY in personal-sign because its the only one that works with detectSIWE.

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (brad-decker) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:174]: I don't think we need this call here, it should only work with personal-sign. Why do we need the isSIWEMessage?

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (filipsekulic) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js:174]: Oh, I thought it's not just for the personal-sign. I'll fix it now.

Comment (filipsekulic) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js]: @brad-decker
You were right, there is no need for the `isSIWEMessage`... I don't know why I put it, thought it might was mandatory for all signatures, because at that moment I was looking into the request data being sent for the personal-sign. Suppose that was the case.
Thanks once again! Hope it's okay now.

**Review by brad-decker — DISMISSED** [review]

**Review by jpuri — DISMISSED** [review]
👍

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (digiwand) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:190]: `securityProviderRequest` may throw an error from metamask-controller.js which could break the flow here. I think we should catch errors here to not allow errors from `securityProviderCheck` to stop the dapp request. 

Current error flow that may occur:
- we call `securityProviderRequest` here in createRPCMethodTrackingMiddleware.js
- `securityProviderRequest` (metamask-controller.js)  calls `securityProviderCheck` to fetch validation
- error occurs when calling `securityProviderCheck`
- we catch the error in metamask-controller.js and then also throw the error again
- error breaks createRPCMethodTrackingMiddleware.js and stops dapp request

**Review by digiwand — CHANGES_REQUESTED** [review]
if the fetch request from `securityProviderCheck` somehow fails, it will stop the dapp request. We will need to handle the error here

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (digiwand) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js]: hey @filipsekulic, I went ahead and merged a solution for the above concern because it seems we're aiming for v10.27 and I was pretty close to a solution after debugging/reviewing. does this look ok to you? [4288143](https://github.com/MetaMask/metamask-extension/pull/17688/commits/428814312a7870ed6c5ede61e4f45370ab28b753)

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (digiwand) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:209]: choosing `console.warn` here since we already log the error in `securityProviderRequest` (metamask-controller.js). maybe there are better ways to go about this. happy to hear other suggestions

File: app/scripts/lib/createRPCMethodTrackingMiddleware.test.js
Comment (digiwand) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.test.js:15]: I think it would be good to add tests for the new `securityProviderRequest` logic in this file

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (digiwand) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js]: cc: @brad-decker @jpuri

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (filipsekulic) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js]: Sounds okay to me.

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (filipsekulic) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js]: It seems fine to me

**Comment by filipsekulic** [comment]
> Some small issues from QA:
> 
> * For transactions, the `ui_customizations` field is appearing on the `Transaction Approved` and `Transaction Rejected` events. Additionally, the field is also included on `Transaction Added` and `Transaction Submitted` events. It make sense to me, but since it is not exactly what it states in the description, I wanted make sure that this is the expected behaviour.
> 
> > Additionally there is a new property called ui_customizations added to Transaction Approved, Transaction Rejected, Signature Approved and Signature Rejected events.
> 
> * For signatures, the `ui_customizations` field is displayed on the `Signature Requested` event, but not on the `Signature Approved` or `Signature Rejected` events. I think this is not the expected behaviour, according to the description

I fixed it. Also, I would appreciate if @brad-decker could take a look once again and confirm your the issues you mentioned.

File: app/scripts/controllers/transactions/index.js
Comment (digiwand) [inline app/scripts/controllers/transactions/index.js:2319]: maybe we want to do something similar to the `transaction_approval_amount_type` property below and only create the key/value if `securityProviderResponse?.flagAsDangerous === 1` is true

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (digiwand) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:196]: similar to https://github.com/MetaMask/metamask-extension/pull/17688/files#r1132887210, maybe we want to not add `ui_customizations` property if empty

Comment (filipsekulic) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js]: Fixed on both places

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (digiwand) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:201]: since on line 193 we have
```          properties.ui_customizations =
            securityProviderResponse?.flagAsDangerous === 1 && [
              'flagged_as_malicious',
            ];
```
it's not guaranteed that `ui_customizations` will be an array. This will cause an error

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (digiwand) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:196]: if `flagAsDanerous !== 1` this will return a boolean 

Screen Shot 2023-03-14 at 3 34 17 PM

**Comment by jpuri** [comment]
@digiwand : has raised some nice points above that should be addressed.

PR looks good otherwise and I can approve once that is done.

File: app/scripts/controllers/transactions/index.test.js
Comment (digiwand) [inline app/scripts/controllers/transactions/index.test.js:1859]: I am thinking that if we are passing `ui_customizations` we should return null rather than false so that we are not introducing a new data type

**Comment by filipsekulic** [comment]
@digiwand @jpuri @brad-decker 
FYI: I have added a solution for [this issue](https://github.com/MetaMask/metamask-extension/issues/18150) in this PR

File: app/scripts/controllers/transactions/index.test.js
Comment (digiwand) [inline app/scripts/controllers/transactions/index.test.js:1744]: should we update this to a number type, rather than a string type, to match with `app/scripts/controllers/transactions/index.js`?

same comment for below instances

**Review by digiwand — DISMISSED** [review]
left one small comment and I think the rest seems good to merge

File: app/scripts/controllers/transactions/index.test.js
Comment (filipsekulic) [reply app/scripts/controllers/transactions/index.test.js]: Good catch! 
I fixed it.

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (filipsekulic) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js]: Good catch @seaona! 👏 Thanks!
Fixed it.

**Review by brad-decker — APPROVED** [review]

**Review by jpuri — APPROVED** [review]

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js
Comment (digiwand) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:177]: upon a second look today, I realized this is an issue. It can splay a string. Not sure how we missed this the first time

Comment (digiwand) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js:177]: working with another engineer to apply a fix for this

**Pull request merged**
